project_forecast_line/models/project_task.py

```diff
@@ -0,0 +1,156 @@
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        # compatibility with fields from project_enterprise
+        for vals in vals_list:
+            if vals.get("planned_date_begin"):
+                vals["forecast_date_planned_start"] = vals["planned_date_begin"]
+            if vals.get("planned_date_end"):
+                vals["forecast_date_planned_end"] = vals["planned_date_end"]
+        tasks = super().create(vals_list)
+        tasks._update_forecast_lines()
+        return tasks
+
+    def _update_forecast_lines_trigger_fields(self):
+        return [
+            "sale_order_line_id",
+            "forecast_role_id",
+            "forecast_date_planned_start",
+            "forecast_date_planned_end",
+            "remaining_hours",
+            "name",
+            "planned_time",
+            "user_id",
+        ]
+
+    def write(self, values):
+        # compatibility with fields from project_enterprise
+        if "planned_date_begin" in values:
+            values["forecast_date_planned_start"] = values["planned_date_begin"]
+        if "planned_date_end" in values:
+            values["forecast_date_planned_end"] = values["planned_date_end"]
+        res = super().write(values)
+        written_fields = list(values.keys())
+        trigger_fields = self._update_forecast_lines_trigger_fields()
+        if any(field in written_fields for field in trigger_fields):
+            self._update_forecast_lines()
+        return res
+
+    @api.onchange("user_id")
+    def onchange_user_id(self):
+        for task in self:
+            if not task.user_id:
+                continue
+            if task.forecast_role_id:
+                continue
+            employee = task.user_id.employee_id
+            if employee.main_role_id:
+                task.forecast_role_id = employee.main_role_id
+                break
+
+    def _update_forecast_lines(self):
+        today = fields.Date.context_today(self)
+        forecast_vals = []
+        ForecastLine = self.env["forecast.line"].sudo()
+        # XXX try to be smarter and only unlink those needing unlinking, update the others
+        ForecastLine.search(
+            [("res_id", "in", self.ids), ("res_model", "=", self._name)]
+        ).unlink()
+        for task in self:
+            if not task.forecast_role_id:
+                _logger.info("skip task %s: no forecast role", task)
+                continue
+            elif task.project_id.project_status:
+                forecast_type = task.project_id.project_status.forecast_line_type
+                if not forecast_type:
+                    _logger.info("skip task %s: no forecast for project state", task)
+                    continue  # closed / cancelled stage
+            elif task.sale_line_id:
+                sale_state = task.sale_line_id.state
+                if sale_state == "cancel":
+                    _logger.info("skip task %s: cancelled sale", task)
+                    continue
```

Comment:
@gurneyalex I added this `continue` here, but it is not there in 15.0 version. In my understanding, it should be there. Am I missing something?